django/db/models/query.py

```diff
@@ -1874,6 +1881,15 @@ def _clone(self):
     def _fetch_all(self):
         if self._result_cache is None:
             self._result_cache = list(self._iterable_class(self))
+            # If we're fetching objects tell them about each other so they can
+            # prefetch for each other during attribute access.
+            if (
+                issubclass(self._iterable_class, ModelIterable)
+                and fetch_missing_fields.get_current_related_strategy() == "peers"
+            ):
+                peers = [weakref.ref(peer) for peer in self._result_cache]
```

Comment:
This should be a `WeakValueDictionary` as in django-auto-prefetch:

https://github.com/tolomea/django-auto-prefetch/blob/0daee9c12f613542cd2d71c8c5d921a63a24869f/src/auto_prefetch/__init__.py#L102

I believe @tolomea made the change to handle the case where a queryset returns duplicate objects, and there was some kind of garbage-collection race condition bug (?) that occurred occasionally. Not sure if [this test](https://github.com/tolomea/django-auto-prefetch/blob/0daee9c12f613542cd2d71c8c5d921a63a24869f/tests/test_basic.py#L149-L171) covers the issue.

Reply:
I think the list of weakrefs is fine. But specific tests would definitely be good.

The problem I had way back when was I used a WeakSet, since Django hashes model instances on PK when a queryset had duplicate entries only one would end up in the WeakSet, the others would fall out, but still have peers set.

My fix was using WeakValueDict keyed off `id(peer)`, but I think list of weak refs also works.